File: 25-reverse-nodes-in-k-group/reverse-nodes-in-k-group.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    pair<ListNode*,ListNode*> reverse(ListNode *head, int k){
        ListNode *prev = NULL;
        ListNode *curr = head;
        int count = 0;
        while(curr && (count < k)){
            ListNode *temp = curr->next;
            curr->next = prev;
            prev = curr;
            curr = temp; 
            count++;
        }
        return {prev, curr};
    }
    ListNode* recur(ListNode *head, int count, int k){
        if(head == NULL || count == 0) return head;
        auto [curr, next] = reverse(head, k);
        head->next = recur(next, count-1, k);
        return curr;
    }
    ListNode* reverseKGroup(ListNode* head, int k) {
        int len = 0;
        ListNode *temp = head;
        while(temp){
            temp = temp->next;
            len++;
        }
        int grps = len/k;
        return recur(head, grps, k);
    }
};

File: 25-reverse-nodes-in-k-group/reverse-nodes-in-k-group.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:

        if k == 0:
            return head

        # Determine length of LL
        length = 0
        curr = head
        while(curr):
            length += 1
            curr = curr.next
        
        n = length // k

        def reverseLL(node):
            prev = None
            curr = node
            count = k
            while(curr and count):
                temp = curr.next
                curr.next = prev
                prev = curr
                curr = temp
                count -= 1
            
            return prev, curr
        

        def recur(node, n):
            if n == 0 or node is None:
                return node
            
            prev, curr = reverseLL(node)

            node.next = recur(curr, n-1)

            return prev
        
        return recur(head, n)
            

        